File: 0101-0200/0159-Longest Substring with At Most Two Distinct Characters/0159-Longest Substring with At Most Two Distinct Characters.py
import collections
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        table = collections.Counter()
        count = start = maxLen = 0
        for i, c in enumerate(s):
            table[c] += 1
            if table[c] == 1:
                count += 1
                while count > 2:
                    table[s[start]] -= 1
                    if table[s[start]] == 0:
                        count -= 1
                    start += 1
            maxLen = max(maxLen, i - start + 1)
        return maxLen

File: 0101-0200/0159-Longest Substring with At Most Two Distinct Characters/0159-Longest Substring with At Most Two Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        vector<int> table(256);
        int count = 0, start = 0, maxLen = 0;
        for (int i = 0; i < s.size(); ++i) {
            if (table[s[i]]++ == 0) {
                ++count;
                while (count > 2) {
                    if (--table[s[start++]] == 0) {
                        --count;
                    }
                }
            }            
            maxLen = max(maxLen, i - start + 1);
        }
        
        return maxLen;        
    }
};
